File: lib/server/query.ts
/**
 * Retrieve the most relevant nearby SideQuest places.
 *
 * Ported from the original agent/query.py.
 */

import { createStore, MossStore } from './store.ts';
import type { SearchResult, Store } from './store.ts';
import { applyContext as applyContextRules } from './context.ts';

export const INDEX_NAME = 'sidequest-places';
const CANDIDATE_COUNT = 100;

/** Metres of walking distance assumed per minute of the caller's radius budget. */
const METRES_PER_RADIUS_MINUTE = 80;
const DEFAULT_RADIUS_MIN = 15;
const MAX_DISTINCT_PLACES = 5;

export interface Constraints {
  radius_min?: number;
  indoor?: boolean;
  lang?: string;
  max_busyness?: number;
  [key: string]: unknown;
}

export interface QueryResult {
  chunks: SearchResult[];
  latency_ms: number;
  warnings: string[];
  user_facts: unknown[];
}

/**
 * The query path is intentionally backed by one long-lived store. In particular,
 * MossStore loads the places index once rather than on every query.
 *
 * The Python version built this at import time. Here it is lazy: Next.js
 * evaluates modules during `next build`, where MOSS_PROJECT_ID is typically
 * absent, and an eager MossStore() would fail the build rather than the request.
 * Module scope still persists across warm invocations, so the index load is
 * amortized exactly as before.
 */
let storePromise: Promise<Store> | null = null;

function getStore(): Promise<Store> {
  storePromise ??= (async () => {
    const store = createStore();
    if (store instanceof MossStore) {
      await store.loadIndex(INDEX_NAME);
    }
    return store;
  })();
  return storePromise;
}

/** Replace the cached store. Tests use this to inject a FakeStore. */
export function setStore(store: Store | null): void {
  storePromise = store ? Promise.resolve(store) : null;
}

/** Return up to five distinct nearby places ranked by semantic relevance. */
export async function query(
  userLat: number,
  userLng: number,
  utterance: string,
  constraints: Constraints | null = null,
  placeId: string | null = null,
): Promise<QueryResult> {
  const [conditioned, contextWarnings] = applyContext(constraints ?? {});
  const filters = nativeFilters(conditioned, placeId);
  const radiusM = Number(conditioned.radius_min ?? DEFAULT_RADIUS_MIN) * METRES_PER_RADIUS_MINUTE;

  let warnings = contextWarnings;
  const startedAt = performance.now();

  let candidates: SearchResult[];
  try {
    const store = await getStore();
    candidates = await store.search(INDEX_NAME, utterance, {
      lat: userLat,
      lng: userLng,
      radiusM,
      filters: Object.keys(filters).length ? filters : null,
      topK: CANDIDATE_COUNT,
    });
  } catch {
    // Fail soft (AGENTS.md): a Moss outage degrades the answer, it does not
    // break the voice path. The caller still gets a well-formed result.
    candidates = [];
    warnings = [...warnings, 'Nearby places are temporarily unavailable.'];
  }

  const latencyMs = performance.now() - startedAt;

  return {
    chunks: distinctPlaces(candidates, conditioned),
    latency_ms: latencyMs,
    warnings,
    user_facts: [],
  };
}

export type ContextHook = (constraints: Constraints) => [Constraints, string[]];

let contextHook: ContextHook | null = null;

/**
 * Override context conditioning. Tests use this in place of the monkeypatching
 * the Python suite did against the module-private `_apply_context`.
 */
export function setContextHook(hook: ContextHook | null): void {
  contextHook = hook;
}

/**
 * Apply context conditioning to the caller's constraints.
 *
 * Fails soft: a throw from the context rules degrades to the unconditioned
 * constraints rather than taking down the query path (AGENTS.md).
 */
function applyContext(constraints: Constraints): [Constraints, string[]] {
  const hook = contextHook ?? applyContextRules;
  try {
    return hook({ ...constraints });
  } catch {
    return [{ ...constraints }, []];
  }
}

/** Use only exact filters at the Store boundary; numeric values are encoded. */
function nativeFilters(
  constraints: Constraints,
  placeId: string | null,
): Record<string, unknown> {
  const filters: Record<string, unknown> = {};
  for (const key of ['indoor', 'lang'] as const) {
    if (key in constraints && constraints[key] !== undefined) {
      filters[key] = constraints[key];
    }
  }
  if (placeId !== null) filters.place_id = placeId;
  return filters;
}

/** Apply encoded numeric constraints and retain the best chunk for each place. */
function distinctPlaces(candidates: SearchResult[], constraints: Constraints): SearchResult[] {
  const maxBusyness = constraints.max_busyness;
  const chunks: SearchResult[] = [];
  const seenPlaceIds = new Set<string>();

  for (const candidate of candidates) {
    const metadata = candidate.metadata ?? {};

    if (maxBusyness !== undefined && asNumber(metadata.busyness_pct) > Number(maxBusyness)) {
      continue;
    }

    const placeId = metadata.place_id;
    if (typeof placeId !== 'string' || !placeId || seenPlaceIds.has(placeId)) continue;

    seenPlaceIds.add(placeId);
    chunks.push(candidate);
    if (chunks.length === MAX_DISTINCT_PLACES) break;
  }

  return chunks;
}

/**
 * Coerce a metadata value to a number, treating anything unparseable as
 * Infinity so a malformed record fails the busyness ceiling rather than
 * slipping past it.
 */
function asNumber(value: unknown): number {
  // Number() is deliberately not used bare: it maps null, "", and [] to 0,
  // which would let a record with a missing busyness_pct slip under the
  // ceiling. Python's float() raised on all three, excluding them instead.
  if (value === null || value === undefined) return Number.POSITIVE_INFINITY;
  if (typeof value !== 'number' && typeof value !== 'string') {
    return Number.POSITIVE_INFINITY;
  }
  if (typeof value === 'string' && value.trim() === '') return Number.POSITIVE_INFINITY;

  const parsed = Number(value);
  return Number.isNaN(parsed) ? Number.POSITIVE_INFINITY : parsed;
}

File: agent/query.py
"""Retrieve the most relevant nearby SideQuest places."""

from __future__ import annotations

import argparse
import json
import time

from lib.store import MossStore, Store


INDEX_NAME = "sidequest-places"
_CANDIDATE_COUNT = 100

# The query path is intentionally backed by one long-lived store.  In
# particular, MossStore loads the places index here rather than on every query.
STORE = Store.create()
if isinstance(STORE, MossStore):
    STORE.load_index(INDEX_NAME)


def query(
    user_lat: float,
    user_lng: float,
    utterance: str,
    constraints: dict | None = None,
    place_id: str | None = None,
) -> dict:
    """Return up to five distinct nearby places ranked by semantic relevance."""
    conditioned_constraints, warnings = _apply_context(constraints or {})
    filters = _native_filters(conditioned_constraints, place_id)
    radius_m = float(conditioned_constraints.get("radius_min", 15)) * 80

    started_at = time.perf_counter()
    try:
        candidates = STORE.search(
            INDEX_NAME,
            utterance,
            lat=user_lat,
            lng=user_lng,
            radius_m=radius_m,
            filters=filters or None,
            top_k=_CANDIDATE_COUNT,
        )
    except Exception:
        candidates = []
        warnings = [*warnings, "Nearby places are temporarily unavailable."]
    latency_ms = (time.perf_counter() - started_at) * 1000

    return {
        "chunks": _distinct_places(candidates, conditioned_constraints),
        "latency_ms": latency_ms,
        "warnings": warnings,
        "user_facts": [],
    }


def _apply_context(constraints: dict) -> tuple[dict, list[str]]:
    """Use context conditioning when its track is available to this checkout."""
    try:
        from agent.context import apply_context
    except ModuleNotFoundError as error:
        if error.name != "agent.context":
            raise
        return dict(constraints), []
    return apply_context(dict(constraints))


def _native_filters(constraints: dict, place_id: str | None) -> dict:
    """Use only exact filters at the Store boundary; numeric values are encoded."""
    filters = {
        key: constraints[key]
        for key in ("indoor", "lang")
        if key in constraints
    }
    if place_id is not None:
        filters["place_id"] = place_id
    return filters


def _distinct_places(candidates: list[dict], constraints: dict) -> list[dict]:
    """Apply encoded numeric constraints and retain the best chunk for each place."""
    max_busyness = constraints.get("max_busyness")
    chunks: list[dict] = []
    seen_place_ids: set[str] = set()
    for candidate in candidates:
        metadata = candidate.get("metadata", {})
        if (
            max_busyness is not None
            and _as_number(metadata.get("busyness_pct")) > float(max_busyness)
        ):
            continue
        resolved_place_id = metadata.get("place_id")
        if not resolved_place_id or resolved_place_id in seen_place_ids:
            continue
        seen_place_ids.add(resolved_place_id)
        chunks.append(candidate)
        if len(chunks) == 5:
            break
    return chunks


def _as_number(value: object) -> float:
    try:
        return float(value)
    except (TypeError, ValueError):
        return float("inf")


def main() -> None:
    parser = argparse.ArgumentParser(description="Query nearby SideQuest places.")
    parser.add_argument("utterance")
    args = parser.parse_args()
    print(json.dumps(query(37.7793, -122.3931, args.utterance), indent=2))


if __name__ == "__main__":
    main()
